### PR TITLE
GG-20879 .NET: specify expiry policy when creating cache using thin client

### PR DESCRIPTION
**Pull request by ashapkin**

### CODE REVIEW

File: modules/platforms/dotnet/Apache.Ignite.Core/Impl/Client/Cache/CacheClient.cs [hunk 1]
```diff
@@ -41,6 +43,22 @@ namespace Apache.Ignite.Core.Impl.Client.Cache
...
+        /// <summary>
+        /// Additional flag values for cache operations.
+        /// </summary>
+        private enum ClientCacheRequestFlag : byte
+        {
+            /// <summary>
+            /// No flags
+            /// </summary>
+            None = 0,
+            // 1 << 0 is reserved for Java thin client.
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Impl/Client/Cache/CacheClient.cs]: Let's put `WithKeepBinary` here anyway, for consistency.

File: modules/platforms/dotnet/Apache.Ignite.Core/Impl/Client/Cache/CacheClient.cs [hunk 2]
```diff
@@ -540,6 +563,14 @@ CacheConfiguration ICacheInternal.GetConfiguration()
             return new CacheClient<TK1, TV1>(_ignite, _name, true);
         }
 
+        /** <inheritDoc /> */
+        public ICacheClient<TK, TV> WithExpiryPolicy(IExpiryPolicy plc)
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Impl/Client/Cache/CacheClient.cs]: Please update `WithKeepBinary` method to pass along `_expiryPolicy`, and add tests to make sure that `WithKeepBinary().WithExpiryPolicy()` works, and vice versa.

Comment (ashapkin) [reply modules/platforms/dotnet/Apache.Ignite.Core/Impl/Client/Cache/CacheClient.cs]: Done

File: modules/core/src/main/java/org/apache/ignite/internal/processors/platform/utils/PlatformConfigurationUtils.java [hunk 3]
```diff
@@ -289,7 +289,7 @@ public static CacheConfiguration readCacheConfiguration(BinaryRawReaderEx in, Cl
      * @param in Reader.
      * @return Expiry policy factory.
      */
-    private static Factory<? extends ExpiryPolicy> readExpiryPolicyFactory(BinaryRawReader in) {
+    public static Factory<? extends ExpiryPolicy> readExpiryPolicyFactory(BinaryRawReader in) {
```
Comment (ptupitsyn) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/platform/utils/PlatformConfigurationUtils.java]: This change and the one and below seem to be unnecessary (no usages outside of this class)

File: modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/cache/ClientCacheRequest.java [hunk 4]
```diff
@@ -69,6 +82,15 @@ protected boolean isKeepBinary() {
         return (flags & FLAG_KEEP_BINARY) == FLAG_KEEP_BINARY;
     }
 
+    /**
+     *  Gets a value indicating whether expiry policy is set in this request.
+     *
+     * @return expiry policy flag value.
+     */
+    protected boolean withExpiryPolicy() {
```
Comment (ptupitsyn) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/cache/ClientCacheRequest.java]: This method is not used outside of this class. I would inline it in the constructor, and use `expiryPolicy != null` check in `rawCache` to simplify.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/cache/ClientCacheRequest.java [hunk 5]
```diff
@@ -80,7 +102,10 @@ protected IgniteCache rawCache(ClientConnectionContext ctx) {
 
         String cacheName = cacheDesc.cacheName();
 
-        return ctx.kernalContext().grid().cache(cacheName);
+        IgniteCache<Object, Object> cache = ctx.kernalContext().grid().cache(cacheName);
+        if (withExpiryPolicy())
+            cache = cache.withExpiryPolicy(expiryPolicy);
+        return cache;
```
Comment (ptupitsyn) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/cache/ClientCacheRequest.java]: Blank line is required after the end of the `if` block

File: modules/platforms/dotnet/Apache.Ignite.Core.Tests/Client/Cache/CacheTest.cs [hunk 6]
```diff
@@ -916,6 +917,151 @@ public void TestCacheNames()
...
+            var expiryPolicy = new ExpiryPolicy(TimeSpan.FromMilliseconds(100), null, null);
+            var cache = GetClientCache<int>();
+            var cacheWithExpiryPolicy = cache.WithExpiryPolicy(expiryPolicy);
+
+            cache.Put(val, val);
+            cacheWithExpiryPolicy.Put(val + 1, val);
+
+            Thread.Sleep(200);
+            
+            Assert.IsTrue(cache.ContainsKey(val));
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core.Tests/Client/Cache/CacheTest.cs]: Let's add cross-checks as well:
```
Assert.IsTrue(cacheWithExpiryPolicy.ContainsKey(val));
Assert.IsFalse(cache.ContainsKey(val + 1);
```

File: modules/platforms/dotnet/Apache.Ignite.Core.Tests/Client/Cache/CacheTest.cs [hunk 7]
```diff
@@ -916,6 +917,151 @@ public void TestCacheNames()
...
+            cacheWithExpiryPolicy.Put(val, val);
+
+            // Initially added value is the same.
+            Assert.IsTrue(cacheWithExpiryPolicy.ContainsKey(val));
+
+            // Wait for an expiration.
+            Thread.Sleep(200);
+
+            // Expiry policies should be applied, no cache item exists.
+            Assert.IsFalse(cacheWithExpiryPolicy.ContainsKey(val));
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core.Tests/Client/Cache/CacheTest.cs]: Here and below - let's add a check for original cache too (without expiry policy).

File: modules/platforms/dotnet/Apache.Ignite.Core/Impl/Client/Cache/CacheClient.cs [hunk 8]
```diff
@@ -41,6 +43,28 @@ namespace Apache.Ignite.Core.Impl.Client.Cache
...
+        /// <summary>
+        /// Additional flag values for cache operations.
+        /// </summary>
+        private enum ClientCacheRequestFlag : byte
+        {
+            /// <summary>
+            /// No flags
+            /// </summary>
+            None = 0,
+            /// <summary>
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Impl/Client/Cache/CacheClient.cs]: Blank line missing before XMLDoc

File: modules/platforms/dotnet/Apache.Ignite.Core/Client/Cache/ICacheClient.cs [hunk 9]
```diff
@@ -422,5 +423,16 @@ public interface ICacheClient<TK, TV>
         /// <typeparam name="TV1">Value type in binary mode.</typeparam>
         /// <returns>Cache instance with binary mode enabled.</returns>
         ICacheClient<TK1, TV1> WithKeepBinary<TK1, TV1>();
+
+        /// <summary>
+        /// Returns cache with the specified expired policy set. This policy will be used for each operation
+        /// invoked on the returned cache.
+        /// <para />
+        /// Expiry durations for each operation are calculated only once and then used as constants. Please
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Client/Cache/ICacheClient.cs]: This is not correct, we actually call `IExpiryPolicy` members on every operation (in contrast with thick mode)

File: modules/core/src/main/java/org/apache/ignite/internal/processors/platform/utils/PlatformConfigurationUtils.java (hunk 3 above)
Comment (ashapkin) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/platform/utils/PlatformConfigurationUtils.java]: I've made this method public in order to reuse it in ClientCacheConfigurationSerializer.java
when we need to parse a new client config section:
```
     case EXPIRY_POLICY: 
                 .....
```
So, I think it's ok

File: modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/cache/ClientCacheRequest.java (hunk 4 above)
Comment (ashapkin) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/cache/ClientCacheRequest.java]: Thanks, converted to private. 
I think the explicit method declaration is a kind of self-documented approach, I would prefer to keep it as is. Thoughts?

Comment (ptupitsyn) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/cache/ClientCacheRequest.java]: No objections

**Pull request merged**